docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,453 @@
+the security subsystem is bootstrapped through the use of Consul health checks
+to gate the initialization of security services.
+This solution worked for the most part,
+but suffers a critical flaw:
+Consul is used in the EdgeX framework for service location,
+and Consul itself is not configured in a secure mode.
+We need a bootstrapping mechanism that ensures that all services,
+including Conul, are brought up in a secure configuration.
+It is impossible,
+once certain services have been started (such as Consul),
+to go back and secure them on-the-fly.
+
+Secure bootstrapping is needed for any installation of EdgeX,
+but the problem is more profound for a container-based deployment of EdgeX
+because a docker-compose script does not have an explicit installation phase.
+The information in this ADR is generally applicable to the install phase of an EdgeX snap,
+or any manual installation of EdgeX.
+
+A security-disabled installation of EdgeX is not required to follow this guidance.
+
+### Items requiring secure bootstrapping
+
+#### Secrets seed
+
+The secrets seed is a hardware-protected secret value that is used for key derivation.
+The **seed** can be used as input key material to a key deriviation function
+that generates an unlimited number of derived keys from an initial secret,
+all of which appear to be random.
+
+Derived keys can be used to protect bootstrapping secrets
+that cannot otherwise be protected by the EdgeX secret store.
+
+#### Consul
+
+Consul is a dynamic configuration database that is initially seeded from static configuration files.
+Putting configuration "online" exposes it to unauthorized modification and disclosure
+and requires introducing compensating controls.
+Most at risk is service location:
+information regarding the host and port where an EdgeX service is available.
+To a lesser extent,
+Consul holds general service configuration information
+and is the fallback location for the EdgeX secret store if Vault is not used.
+Consul also provides for service health checks.
+
+The elements required to secure Consul are:
+
+- For neighbor discovery: An **encryption key** used by Serf;
+  this key must be provided the first time that Consul is started.
+  This key prevents rogue Consul instances from joining the cluster.
+  This key must also be provided to any new Consul agents that are added to the system.
+  This key can be a 32-byte derived key or a random key.
+- For TLS: a **CA certificate**, **TLS certificate (chain)**, and **TLS private key**.
+
+For integration with Vault,
+some initial secrets must be created:
+
+- A **management token** that can be passed to Vault's Consul secrets engine,
+  to enable a Vault API to be utilized to create additional per-microservice Consul tokens.
+- An **ACL token** that allows Vault to read/write a portion of Consul's key space as the backing store for Vault.
+  (Intent is to switch to the Vault raft storage backend to avoid Consul dependency.)
```

Comment:
what problem is this solving and what downsides are involved?

Past experiences with raft have not been pleasant. Of specific concern is does this still permit vault to distribute secrets via consul?

Reply:
also is the vault->consul acl token something that currently happens

Reply:
The proposal here is to have Vault distribute secrets via Vault.  If we turn on Consul authentication, then Vault can no longer use Consul as it storage back-end unless it has a Consul token.  The only reason it works now is because Consul does not require authentication.  To fix his means creating Consul secrets outside of Vault and using these external secrets to configure Vault itself.  According to the "Creation and Distribution of Secrets" ADR we want Vault to be the system of reference of secrets, and configuring the system in the Vault-depends-on-Consul manner puts Consul in the number-one position versus Vault in the number-one position. Alternatively, I would have to generate a parallel token-granting infrastructure for Consul that looks almost like the one for Vault, rather than relying on Vault's built-in Consul secrets engine.

I have a hard time accepting the criticism that raft is problematic whereas Consul is not, given that Consul is based on Raft: https://www.consul.io/docs/internals/consensus.html.

Reply:
Agree with "vault as the system of reference" does this ADR still allow for consul as a distributor / arrbitrator of said secrets.

Reply:
This ADR as-written uses Vault for secret management in the security-enabled configuration.  Consul was not designed as a secret management engine.  IMO in the security-enabled configuration secrets should not be passed through Consul.

Reply:
Should consul not be left in per the secret creation and distribution ADR?

Without consul to distribute secrets does one end up reinventing that wheel including but not limited to ACL, tokens, encryption, auditing?

Ping me on slack if I'm way off base but the addition of `envconsul` as a second alternative and `consul-template` as the third option gave me the impression this ADR would need to leave them behind or at least leave them behind until alternatives are implemented

Reply:
> Ping me on slack if I'm way off base but the addition of `envconsul` as a second alternative and `consul-template` as the third option gave me the impression this ADR would need to leave them behind or at least leave them behind until alternatives are implemented

* envconsul - Documentation says that it can pull a secret from Vault and put it into the environment.
* consul-template - Documentation says that this tool also supports pulling a secret from vault.

So despite the fact that they have "consul" in the name, they do both work with Vault and thus can be used for implementation of this ADR.

> Without consul to distribute secrets does one end up reinventing that wheel including but not limited to ACL, tokens, encryption, auditing?

The non-secure bootstrap would simply use secrets from configuration files or none at all, via conditional logic (EDGEX_SECURITY_SECRET_STORE="false").